### ISSUE TITLE
Angular - Form Validation (Duplicate call function custom validate)

### ISSUE THREAD

**Issue by laptransang** [angular/angular#21500]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

I am testing to make form-validation custom and i found something unreasonable, when the first init component new FormGroup, i pass the function name like customValidator to custom validate and i console.log to test it action, and i see new FormGroup is call three time action like this:

the command console.log is console.log('customValidator', control);

customValidator FormControl {validator: ƒ, asyncValidator: null, _onCollectionChange: ƒ, pristine: true, touched: false, …}
hero-form-reactive.component.ts:31 customValidator FormControl {validator: ƒ, asyncValidator: null, _onCollectionChange: ƒ, pristine: true, touched: false, …}
hero-form-reactive.component.ts:31 customValidator FormControl {validator: ƒ, asyncValidator: null, _onCollectionChange: ƒ, pristine: true, touched: false, …}

## Expected behavior

Because it call three time lead to when you call api to async it will duplicate request, i can't fix temporary with debounce lodash, but i think it shound be call one time when init

[Link to Repro](https://plnkr.co/edit/HOOnul9tqUagPB1P4E0l?p=preview)

When you open Developer Tools you will see it call three time actions in custom function validate.

## Environment

<pre><code>
Angular version: 5.1.0

Browser:
- [x] Chrome (desktop) version  63.0.3239.132 (Official Build) (64-bit)
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: v8.9.3
- Platform:  Mac

Others:

</code></pre>

**Comment by maximelafarie** [comment]
Error in your plunkr.
`import { HeroFormTemplateComponent } from './template/hero-form-template.component';` returns a 404 given that it doesn't exist.

**Comment by laptransang** [comment]
Yep @maximelafarie , i forgot to remove import in app.module.ts, now it woking, and in debug you will see duplicate call.

**Comment by kara** [comment]
@laptransang It's expected that the validator would be called at least twice: 

- once when the FormControl is created (so its status is correct with relation to its initial value) 
- and once when the FormControlName directive is created (so the status can be updated if there are validator directives defined in the template)

It looks like the validator is called a third time as a part of FormGroupDirective's global update of the tree. We may be able to prevent this third time if we move some things around, but will need further testing.

**Comment by laptransang** [comment]
Yep thanks for your response @kara, I think twice is more reasonable, i will use debounce lodash decorator to fix asynchronous call validate .

**Comment by guillep2k** [comment]
Issue #20052 provides more test cases for this (it was closed by mistake, apparently).

**Comment by jarodsmk** [comment]
Any updates on this?

I'm going to implement a `Subject` to `debounce` the queries coming through as a solution for now.

**Comment by rubmz** [comment]
Can't believe this is the behavior ... It makes it impossible to validate two or more fields asynchronously. The "asyncValidator" keeps firing with each change to the control, which makes it very complex to return a new observable (that depends on other controls/values) each time. Shouldn't an asyncValidator be called just once and return a custom observable that shall observe  control.valueChanges independently? Otherwise ... it is required to return a closed observable (using take(1) or another way) which cannot wait freely for changes in other controls/data (asynchronously).
